README.rst
Python Client for Cloud AutoML API
==================================

|stable| |pypi| |versions|

`Cloud AutoML API`_: 

- `Client Library Documentation`_
- `Product Documentation`_

.. |stable| image:: https://img.shields.io/badge/support-stable-gold.svg
   :target: https://github.com/googleapis/google-cloud-python/blob/main/README.rst#stability-levels
.. |pypi| image:: https://img.shields.io/pypi/v/google-cloud-automl.svg
   :target: https://pypi.org/project/google-cloud-automl/
.. |versions| image:: https://img.shields.io/pypi/pyversions/google-cloud-automl.svg
   :target: https://pypi.org/project/google-cloud-automl/
.. _Cloud AutoML API: https://cloud.google.com/automl/docs/
.. _Client Library Documentation: https://cloud.google.com/python/docs/reference/automl/latest
.. _Product Documentation:  https://cloud.google.com/automl/docs/

Quick Start
-----------

In order to use this library, you first need to go through the following steps:

1. `Select or create a Cloud Platform project.`_
2. `Enable billing for your project.`_
3. `Enable the Cloud AutoML API.`_
4. `Setup Authentication.`_

.. _Select or create a Cloud Platform project.: https://console.cloud.google.com/project
.. _Enable billing for your project.: https://cloud.google.com/billing/docs/how-to/modify-project#enable_billing_for_a_project
.. _Enable the Cloud AutoML API.:  https://cloud.google.com/automl/docs/
.. _Setup Authentication.: https://googleapis.dev/python/google-api-core/latest/auth.html

Installation
~~~~~~~~~~~~

Install this library in a `virtualenv`_ using pip. `virtualenv`_ is a tool to
create isolated Python environments. The basic problem it addresses is one of
dependencies and versions, and indirectly permissions.

With `virtualenv`_, it's possible to install this library without needing system
install permissions, and without clashing with the installed system
dependencies.

.. _`virtualenv`: https://virtualenv.pypa.io/en/latest/


Code samples and snippets
~~~~~~~~~~~~~~~~~~~~~~~~~

Code samples and snippets live in the `samples/` folder.


Supported Python Versions
^^^^^^^^^^^^^^^^^^^^^^^^^
Our client libraries are compatible with all current `active`_ and `maintenance`_ versions of
Python.

Python >= 3.7

.. _active: https://devguide.python.org/devcycle/#in-development-main-branch
.. _maintenance: https://devguide.python.org/devcycle/#maintenance-branches

Unsupported Python Versions
^^^^^^^^^^^^^^^^^^^^^^^^^^^
Python <= 3.6

If you are using an `end-of-life`_
version of Python, we recommend that you update as soon as possible to an actively supported version.

.. _end-of-life: https://devguide.python.org/devcycle/#end-of-life-branches

Mac/Linux
^^^^^^^^^

.. code-block:: console

    pip install virtualenv
    virtualenv <your-env>
    source <your-env>/bin/activate
    <your-env>/bin/pip install google-cloud-automl


Windows
^^^^^^^

.. code-block:: console

    pip install virtualenv
    virtualenv <your-env>
    <your-env>\Scripts\activate
    <your-env>\Scripts\pip.exe install google-cloud-automl

Next Steps
~~~~~~~~~~

-  Read the `Client Library Documentation`_ for Cloud AutoML API
   to see other available methods on the client.
-  Read the `Cloud AutoML API Product documentation`_ to learn
   more about the product and see How-to Guides.
-  View this `README`_ to see the full list of Cloud
   APIs that we cover.

.. _Cloud AutoML API Product documentation:  https://cloud.google.com/automl/docs/
.. _README: https://github.com/googleapis/google-cloud-python/blob/main/README.rst


CHANGELOG.md
# Changelog

[PyPI History][1]

[1]: https://pypi.org/project/google-cloud-automl/#history

## [2.11.3](https://github.com/googleapis/python-automl/compare/v2.11.2...v2.11.3) (2023-10-16)


### Documentation

* Minor formatting ([e146232](https://github.com/googleapis/python-automl/commit/e1462325c948e800edc93ba45f437d3e66cb5556))

## [2.11.2](https://github.com/googleapis/python-automl/compare/v2.11.1...v2.11.2) (2023-07-04)


### Bug Fixes

* Add async context manager return types ([#490](https://github.com/googleapis/python-automl/issues/490)) ([0fd78dc](https://github.com/googleapis/python-automl/commit/0fd78dc06e133f2fc6a3e41373ea47f4b4ea898d))

## [2.11.1](https://github.com/googleapis/python-automl/compare/v2.11.0...v2.11.1) (2023-03-23)


### Documentation

* Fix formatting of request arg in docstring ([#481](https://github.com/googleapis/python-automl/issues/481)) ([6e78170](https://github.com/googleapis/python-automl/commit/6e78170a97f44583522d9fe0aabd2cb1ea452f7b))

## [2.11.0](https://github.com/googleapis/python-automl/compare/v2.10.1...v2.11.0) (2023-02-28)


### Features

* Enable "rest" transport in Python for services supporting numeric enums ([#472](https://github.com/googleapis/python-automl/issues/472)) ([8d76d92](https://github.com/googleapis/python-automl/commit/8d76d922c48430cb0709cc35e9c10b92536f57c4))

## [2.10.1](https://github.com/googleapis/python-automl/compare/v2.10.0...v2.10.1) (2023-01-20)


### Bug Fixes

* Add context manager return types ([c81ac42](https://github.com/googleapis/python-automl/commit/c81ac42954b261e92996556f2e0a50c9954bb99f))


### Documentation

* Add documentation for enums ([c81ac42](https://github.com/googleapis/python-automl/commit/c81ac42954b261e92996556f2e0a50c9954bb99f))

## [2.10.0](https://github.com/googleapis/python-automl/compare/v2.9.0...v2.10.0) (2023-01-10)


### Features

* Add support for python 3.11 ([#461](https://github.com/googleapis/python-automl/issues/461)) ([06d11c6](https://github.com/googleapis/python-automl/commit/06d11c6084cd809edb8927d914d3aab15023e606))

## [2.9.0](https://github.com/googleapis/python-automl/compare/v2.8.3...v2.9.0) (2022-12-15)


### Features

* Add support for `google.cloud.automl.__version__` ([ea81cb4](https://github.com/googleapis/python-automl/commit/ea81cb4cf5dedd3b143c502261d978dac590cc20))
* Add typing to proto.Message based class attributes ([ea81cb4](https://github.com/googleapis/python-automl/commit/ea81cb4cf5dedd3b143c502261d978dac590cc20))


### Bug Fixes

* Add dict typing for client_options ([ea81cb4](https://github.com/googleapis/python-automl/commit/ea81cb4cf5dedd3b143c502261d978dac590cc20))
* **deps:** Require google-api-core &gt;=1.34.0, >=2.11.0  ([ad2a2d7](https://github.com/googleapis/python-automl/commit/ad2a2d7806593d5f1f19a62562dbf110504b4097))
* Drop usage of pkg_resources ([ad2a2d7](https://github.com/googleapis/python-automl/commit/ad2a2d7806593d5f1f19a62562dbf110504b4097))
* Fix timeout default values ([ad2a2d7](https://github.com/googleapis/python-automl/commit/ad2a2d7806593d5f1f19a62562dbf110504b4097))


### Documentation

* **samples:** Snippetgen handling of repeated enum field ([ea81cb4](https://github.com/googleapis/python-automl/commit/ea81cb4cf5dedd3b143c502261d978dac590cc20))
* **samples:** Snippetgen should call await on the operation coroutine before calling result ([ad2a2d7](https://github.com/googleapis/python-automl/commit/ad2a2d7806593d5f1f19a62562dbf110504b4097))

## [2.8.3](https://github.com/googleapis/python-automl/compare/v2.8.2...v2.8.3) (2022-10-07)


### Bug Fixes

* **deps:** Allow protobuf 3.19.5 ([#444](https://github.com/googleapis/python-automl/issues/444)) ([501d8bf](https://github.com/googleapis/python-automl/commit/501d8bf4c14609a3d55b3dd16afc4e327bdec37d))

## [2.8.2](https://github.com/googleapis/python-automl/compare/v2.8.1...v2.8.2) (2022-10-03)


### Bug Fixes

* **deps:** Require protobuf &gt;= 3.20.2 ([#439](https://github.com/googleapis/python-automl/issues/439)) ([33960ca](https://github.com/googleapis/python-automl/commit/33960ca1facc2f8c3171c51d9b16df4e78ecf922))

## [2.8.1](https://github.com/googleapis/python-automl/compare/v2.8.0...v2.8.1) (2022-08-11)


### Bug Fixes

* **deps:** allow protobuf < 5.0.0 ([#417](https://github.com/googleapis/python-automl/issues/417)) ([cbb1fce](https://github.com/googleapis/python-automl/commit/cbb1fce19f82a8e3e0291ef4856bf28b19d9b198))
* **deps:** require proto-plus >= 1.22.0 ([cbb1fce](https://github.com/googleapis/python-automl/commit/cbb1fce19f82a8e3e0291ef4856bf28b19d9b198))

## [2.8.0](https://github.com/googleapis/python-automl/compare/v2.7.3...v2.8.0) (2022-07-19)


### Features

* add audience parameter ([51f21e1](https://github.com/googleapis/python-automl/commit/51f21e1afd39e8c9a159908b654bcabc73226b94))


### Bug Fixes

* **deps:** require google-api-core>=1.32.0,>=2.8.0 ([#403](https://github.com/googleapis/python-automl/issues/403)) ([51f21e1](https://github.com/googleapis/python-automl/commit/51f21e1afd39e8c9a159908b654bcabc73226b94))
* require python 3.7+ ([#405](https://github.com/googleapis/python-automl/issues/405)) ([8d762e6](https://github.com/googleapis/python-automl/commit/8d762e6490e41d279d38605adca710bb21994808))

## [2.7.3](https://github.com/googleapis/python-automl/compare/v2.7.2...v2.7.3) (2022-06-07)


### Bug Fixes

* **deps:** require protobuf <4.0.0dev ([#391](https://github.com/googleapis/python-automl/issues/391)) ([30cfcb2](https://github.com/googleapis/python-automl/commit/30cfcb2acd5ed43cdbd74610a87bde7eb23dec03))


### Documentation

* fix changelog header to consistent size ([#392](https://github.com/googleapis/python-automl/issues/392)) ([18952b3](https://github.com/googleapis/python-automl/commit/18952b306f91e49b89a3cb3b4022b0b6a8ed4eac))

## [2.7.2](https://github.com/googleapis/python-automl/compare/v2.7.1...v2.7.2) (2022-03-06)


### Bug Fixes


owl-bot-staging/v1/noxfile.py
# -*- coding: utf-8 -*-
# Copyright 2023 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
import os
import pathlib
import shutil
import subprocess
import sys


import nox  # type: ignore

ALL_PYTHON = [
    "3.7",
    "3.8",
    "3.9",
    "3.10",
    "3.11",
]

CURRENT_DIRECTORY = pathlib.Path(__file__).parent.absolute()

LOWER_BOUND_CONSTRAINTS_FILE = CURRENT_DIRECTORY / "constraints.txt"
PACKAGE_NAME = subprocess.check_output([sys.executable, "setup.py", "--name"], encoding="utf-8")

BLACK_VERSION = "black==22.3.0"
BLACK_PATHS = ["docs", "google", "tests", "samples", "noxfile.py", "setup.py"]
DEFAULT_PYTHON_VERSION = "3.11"

nox.sessions = [
    "unit",
    "cover",
    "mypy",
    "check_lower_bounds"
    # exclude update_lower_bounds from default
    "docs",
    "blacken",
    "lint",
    "lint_setup_py",
]

@nox.session(python=ALL_PYTHON)
def unit(session):
    """Run the unit test suite."""

    session.install('coverage', 'pytest', 'pytest-cov', 'pytest-asyncio', 'asyncmock; python_version < "3.8"')
    session.install('-e', '.')

    session.run(
        'py.test',
        '--quiet',
        '--cov=google/cloud/automl_v1/',
        '--cov=tests/',
        '--cov-config=.coveragerc',
        '--cov-report=term',
        '--cov-report=html',
        os.path.join('tests', 'unit', ''.join(session.posargs))
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def cover(session):
    """Run the final coverage report.
    This outputs the coverage report aggregating coverage from the unit
    test runs (not system test runs), and then erases coverage data.
    """
    session.install("coverage", "pytest-cov")
    session.run("coverage", "report", "--show-missing", "--fail-under=100")

    session.run("coverage", "erase")


@nox.session(python=ALL_PYTHON)
def mypy(session):
    """Run the type checker."""
    session.install(
        'mypy',
        'types-requests',
        'types-protobuf'
    )
    session.install('.')
    session.run(
        'mypy',
        '--explicit-package-bases',
        'google',
    )


@nox.session
def update_lower_bounds(session):
    """Update lower bounds in constraints.txt to match setup.py"""
    session.install('google-cloud-testutils')
    session.install('.')

    session.run(
        'lower-bound-checker',
        'update',
        '--package-name',
        PACKAGE_NAME,
        '--constraints-file',
        str(LOWER_BOUND_CONSTRAINTS_FILE),
    )


@nox.session
def check_lower_bounds(session):
    """Check lower bounds in setup.py are reflected in constraints file"""
    session.install('google-cloud-testutils')
    session.install('.')

    session.run(
        'lower-bound-checker',
        'check',
        '--package-name',
        PACKAGE_NAME,
        '--constraints-file',
        str(LOWER_BOUND_CONSTRAINTS_FILE),
    )

@nox.session(python=DEFAULT_PYTHON_VERSION)
def docs(session):
    """Build the docs for this library."""

    session.install("-e", ".")
    session.install("sphinx==4.0.1", "alabaster", "recommonmark")

    shutil.rmtree(os.path.join("docs", "_build"), ignore_errors=True)
    session.run(
        "sphinx-build",
        "-W",  # warnings as errors
        "-T",  # show full traceback on exception
        "-N",  # no colors
        "-b",
        "html",
        "-d",
        os.path.join("docs", "_build", "doctrees", ""),
        os.path.join("docs", ""),
        os.path.join("docs", "_build", "html", ""),
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def lint(session):
    """Run linters.

    Returns a failure if the linters find linting errors or sufficiently
    serious code quality issues.
    """
    session.install("flake8", BLACK_VERSION)
    session.run(
        "black",
        "--check",
        *BLACK_PATHS,
    )
    session.run("flake8", "google", "tests", "samples")


@nox.session(python=DEFAULT_PYTHON_VERSION)
def blacken(session):
    """Run black. Format code to uniform standard."""
    session.install(BLACK_VERSION)
    session.run(
        "black",
        *BLACK_PATHS,
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def lint_setup_py(session):
    """Verify that setup.py is valid (including RST check)."""
    session.install("docutils", "pygments")
    session.run("python", "setup.py", "check", "--restructuredtext", "--strict")
